# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/libwebrtcglue/TaskQueueWrapper.h
# Commit: 8d80a4ba914c
# Full Hash: 8d80a4ba914c6f81430ba1c55471a8d2b3aac60e
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 2
# Description:
#   Bug 1654112 - Create CallWorkerThread which implements AbstractThread for a TaskQueueWrapper. r=ng
#   
#   Ideally TaskQueueWrapper would implement both webrtc::TaskQueueBase and
#   AbstractThread, but webrtc::TaskQueueBase is not refcounted so implementing both
#   in the same class is not possible.
# ==============================================================================

diff -r e0a2e54e4de8 -r 8d80a4ba914c dom/media/webrtc/libwebrtcglue/TaskQueueWrapper.h
--- a/dom/media/webrtc/libwebrtcglue/TaskQueueWrapper.h	Thu Apr 08 11:31:43 2021 +0200
+++ b/dom/media/webrtc/libwebrtcglue/TaskQueueWrapper.h	Mon Apr 12 14:49:20 2021 +0200
@@ -71,6 +71,19 @@
                                   });
   }
 
+  already_AddRefed<Runnable> CreateTaskRunner(nsCOMPtr<nsIRunnable> aRunnable) {
+    return NS_NewRunnableFunction(
+        "TaskQueueWrapper::CreateTaskRunner",
+        [this, runnable = std::move(aRunnable)]() mutable {
+          CurrentTaskQueueSetter current(this);
+          auto hasShutdown = mHasShutdown.Lock();
+          if (*hasShutdown) {
+            return;
+          }
+          runnable->Run();
+        });
+  }
+
   void PostTask(std::unique_ptr<webrtc::QueuedTask> aTask) override {
     MOZ_ALWAYS_SUCCEEDS(
         mTaskQueue->Dispatch(CreateTaskRunner(std::move(aTask))));
@@ -103,14 +116,20 @@
  public:
   SharedThreadPoolWebRtcTaskQueueFactory() {}
 
-  std::unique_ptr<webrtc::TaskQueueBase, webrtc::TaskQueueDeleter>
-  CreateTaskQueue(absl::string_view aName, Priority aPriority) const override {
+  UniquePtr<TaskQueueWrapper> CreateTaskQueueWrapper(absl::string_view aName,
+                                                     Priority aPriority) const {
     // XXX Do something with aPriority
     nsCString name(aName.data(), aName.size());
     auto taskQueue = MakeRefPtr<TaskQueue>(
         GetMediaThreadPool(MediaThreadType::WEBRTC_DECODER), name.get());
+    return MakeUnique<TaskQueueWrapper>(std::move(taskQueue));
+  }
+
+  std::unique_ptr<webrtc::TaskQueueBase, webrtc::TaskQueueDeleter>
+  CreateTaskQueue(absl::string_view aName, Priority aPriority) const override {
     return std::unique_ptr<webrtc::TaskQueueBase, webrtc::TaskQueueDeleter>(
-        new TaskQueueWrapper(std::move(taskQueue)), webrtc::TaskQueueDeleter());
+        CreateTaskQueueWrapper(std::move(aName), aPriority).release(),
+        webrtc::TaskQueueDeleter());
   }
 };
 